File: 0368-largest-divisible-subset/0368-largest-divisible-subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        // Sort the array to ensure smaller numbers come before larger ones
        Arrays.sort(nums);
      
        int n = nums.length;
        // dp[i] represents the length of the largest divisible subset ending at index i
        int[] dp = new int[n];
        Arrays.fill(dp, 1); // Each element forms a subset of size 1 by itself
      
        // Track the index with the maximum subset length
        int maxIndex = 0;
      
        // Build dp array using dynamic programming
        for (int i = 0; i < n; i++) {
            // Check all previous elements
            for (int j = 0; j < i; j++) {
                // If nums[i] is divisible by nums[j], we can extend the subset ending at j
                if (nums[i] % nums[j] == 0) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            // Update the index of maximum subset length
            if (dp[maxIndex] < dp[i]) {
                maxIndex = i;
            }
        }
      
        // Reconstruct the largest divisible subset
        int maxLength = dp[maxIndex];
        List<Integer> result = new ArrayList<>();
      
        // Backtrack from maxIndex to build the result
        for (int i = maxIndex; maxLength > 0; i--) {
            // Check if current element can be part of the subset
            // Conditions: nums[maxIndex] divisible by nums[i] AND dp[i] matches expected length
            if (nums[maxIndex] % nums[i] == 0 && dp[i] == maxLength) {
                result.add(nums[i]);
                maxIndex = i;  // Update maxIndex for next iteration
                maxLength--;   // Decrease expected length
            }
        }
      
        return result;
    }
}

File: 0368-largest-divisible-subset/0368-largest-divisible-subset.ts
/**
 * Finds the largest subset where every pair of elements (a, b) satisfies either a % b = 0 or b % a = 0
 * @param nums - Array of positive integers
 * @returns The largest divisible subset
 */
function largestDivisibleSubset(nums: number[]): number[] {
    // Sort the array in ascending order to ensure smaller elements come before larger ones
    nums.sort((a, b) => a - b);
  
    const arrayLength: number = nums.length;
  
    // Dynamic programming array where dp[i] represents the size of largest subset ending at index i
    const dp: number[] = Array(arrayLength).fill(1);
  
    // Track the index of the element that forms the largest subset
    let maxSubsetIndex: number = 0;

    // Build up the dp array
    for (let currentIndex = 0; currentIndex < arrayLength; ++currentIndex) {
        // Check all previous elements that could form a subset with current element
        for (let previousIndex = 0; previousIndex < currentIndex; ++previousIndex) {
            // If current element is divisible by previous element, they can be in same subset
            if (nums[currentIndex] % nums[previousIndex] === 0) {
                dp[currentIndex] = Math.max(dp[currentIndex], dp[previousIndex] + 1);
            }
        }
      
        // Update the index of maximum subset if current subset is larger
        if (dp[maxSubsetIndex] < dp[currentIndex]) {
            maxSubsetIndex = currentIndex;
        }
    }

    // Reconstruct the actual subset from the dp array
    let remainingElements: number = dp[maxSubsetIndex];
    const result: number[] = [];
  
    // Backtrack from the largest element to build the subset
    for (let index = maxSubsetIndex; remainingElements > 0; --index) {
        // Check if current element can be part of the subset chain
        if (nums[maxSubsetIndex] % nums[index] === 0 && dp[index] === remainingElements) {
            result.push(nums[index]);
            maxSubsetIndex = index;
            --remainingElements;
        }
    }

    return result;
}
